### PR TITLE
osd/PeeringState: do not exclude up from acting_recovery_backfill

### PR DESCRIPTION
**Pull request by xiexingguo**
If we choose a primary that does not belong to the current up set,
and all up peers are still recoverable, then we might end up excluding
some up peer from the acting_recovery_backfill set too due to the
"want size <= pool size" constraint (since https://github.com/ceph/ceph/pull/24035),
as a result of which all up peers might not get recovered in one go.

Fix by falling through any oversized want set to async recovery, which
should be able to handle it nicely.

Fixes: https://tracker.ceph.com/issues/42577
Signed-off-by: xie xingguo <xie.xingguo@zte.com.cn>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/osd/PeeringState.cc [hunk 1]
```diff
@@ -1703,9 +1703,6 @@ void PeeringState::calc_replicated_acting(
       acting_backfill->insert(up_cand);
       ss << " osd." << i << " (up) accepted " << cur_info << std::endl;
     }
-    if (want->size() >= size) {
```
Comment (xiexingguo) [inline src/osd/PeeringState.cc]: @liewegas it's wrong to break here as we still want all up peers go to the acting_backfill set..

**Comment by xiexingguo** [comment]
retest this please

**Comment by dzafman** [comment]
It might be clearer to show this as a revert of c3e2990c39952a1fc4593076c5b2f0dea6cdcdd3 and a commit with the alternate fix.

**Comment by xiexingguo** [comment]
> this as a revert of [c3e2990](https://github.com/ceph/ceph/commit/c3e2990c39952a1fc4593076c5b2f0dea6cdcdd3)

I tried. The conflict of reverting is huge(pg.cc -> peeringstate.cc)  :-(

**Comment by dzafman** [comment]
> > this as a revert of [c3e2990](https://github.com/ceph/ceph/commit/c3e2990c39952a1fc4593076c5b2f0dea6cdcdd3)
> 
> I tried. The conflict of reverting is huge(pg.cc -> peeringstate.cc) :-(

The code moved to src/osd/PeeringState.cc.   You could do the following:
git revert c3e2990,
git reset HEAD src/osd/PG.cc
git checkout -- src/osd/PG.cc
Remove the 3 lines from PeeringState::calc_replicated_acting() in src/osd/PeeringState.cc
git add src/osd/PeeringState.cc
git revert --continue

**Comment by xiexingguo** [comment]
@dzafman Thanks!
@neha-ojha @liewegas Mind taking a look?

**Comment by neha-ojha** [comment]
retest this please

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2019-11-19_09:56:38-rados-wip-kefu2-testing-2019-11-19-1509-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-11-21_08:41:20-rados-wip-kefu2-testing-2019-11-19-1509-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2019-11-23_01:15:32-rados-wip-kefu2-testing-2019-11-23-0140-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2019-11-24_13:33:51-rados-wip-kefu2-testing-2019-11-24-1912-distro-basic-mira/

**Comment by xiexingguo** [comment]
@tchaikov Thanks!

**Pull request merged**